abdiff/cli.py

```diff
@@ -1,27 +1,119 @@
+import json
 import logging
+from collections.abc import Callable
 from datetime import timedelta
 from time import perf_counter
 
 import click
+from click.exceptions import ClickException
 
 from abdiff.config import configure_logger
+from abdiff.core import build_ab_images
+from abdiff.core import init_job as core_init_job
+from abdiff.core.utils import read_job_json
 
 logger = logging.getLogger(__name__)
 
 
-@click.command()
+@click.group(context_settings={"help_option_names": ["-h", "--help"]})
 @click.option(
-    "-v", "--verbose", is_flag=True, help="Pass to log at debug level instead of info"
+    "-v",
+    "--verbose",
+    is_flag=True,
+    help="Pass to log at debug level instead of info.",
 )
-def main(*, verbose: bool) -> None:
-    start_time = perf_counter()
+@click.pass_context
+def main(
+    ctx: click.Context,
+    verbose: bool,  # noqa: FBT001
+) -> None:
+    ctx.ensure_object(dict)
+    ctx.obj["START_TIME"] = perf_counter()
     root_logger = logging.getLogger()
     logger.info(configure_logger(root_logger, verbose=verbose))
     logger.info("Running process")
 
-    # Do things here!
 
-    elapsed_time = perf_counter() - start_time
+@main.result_callback()
+@click.pass_context
+def post_main_group_subcommand(
+    ctx: click.Context,
+    *_args: tuple,
+    **_kwargs: dict,
+) -> None:
     logger.info(
-        "Total time to complete process: %s", str(timedelta(seconds=elapsed_time))
+        "Total elapsed: %s",
+        str(
+            timedelta(seconds=perf_counter() - ctx.obj["START_TIME"]),
+        ),
     )
+
+
+@main.command()
+def ping() -> None:
+    """Debug ping/pong command."""
+    logger.debug("got ping, preparing to pong")
+    click.echo("pong")
```

Comment:
We could, but it is consistent with other projects where a generic ping command (also the first built) follows the main group:

- [geoharvester](https://github.com/MITLibraries/geo-harvester/blob/main/harvester/cli.py#L38-L53)
- [HRQBClient](https://github.com/MITLibraries/hrqb-client/blob/5a35f77455ee4649aa91792f66fbcd280dd35684/hrqb/cli.py#L37-L46)
- [TIM](https://github.com/MITLibraries/timdex-index-manager/blob/b628e094c191e44b40e6602840f74ac3a2e8d125/tim/cli.py#L99-L106)
- [TIMDEX API (ruby code, but same convention of near top)](https://github.com/MITLibraries/timdex/blob/b7f64478dcad709c06020ff4144ee16b7428b61b/app/graphql/types/query_type.rb#L10-L12)

Maybe a more compelling reason is just code churn.  As a CLI command that is likely basically never change, I think there is some reasoning behind keeping it near the top so that, as commands are added, we don't keep moving this code block farther and farther down.

Open to moving if you feel strongly, but it feels like it follows convention for us!